File: models/examples/usage_examples.py
#!/usr/bin/env python3
"""
EQ12 Production Model Client - Usage Examples
Expert-level integration patterns for OpenAI models with EQ12 constraints.

This demonstrates the complete workflow from raw odds → normalized data → parlays
using the task-specific model selection from your expert guide.
"""

# Import our production client
import sys
from pathlib import Path

sys.path.append(str(Path(__file__).parent.parent))
from eq12_client import EQ12Config, EQ12ModelClient


def main():
    """Complete EQ12 model integration example following expert patterns."""

    print("🎰 EQ12 Production Model Client - Usage Examples")
    print("=" * 60)

    # ✅ Initialize with EQ12-specific constraints
    config = EQ12Config(
        allowed_books=["draftkings", "fanduel", "betmgm"],
        min_ev_threshold=0.03,  # 3% minimum edge
        kelly_cap_per_leg=0.025,  # 2.5% Kelly cap per leg
        max_correlation_risk=0.15,  # Low correlation tolerance
        stale_data_threshold_minutes=15,  # Fresh data only
    )

    client = EQ12ModelClient(config)

    # ✅ Example 1: Extract odds from raw sportsbook data
    print("\n📊 STEP 1: Extract & normalize odds (gpt-4o-mini)")
    raw_odds_data = """
    DraftKings NFL Week 5:
    Kansas City Chiefs -3.0 (-110) vs Buffalo Bills +3.0 (-110)
    Total: Over 47.5 (-110), Under 47.5 (-110)
    Last updated: 2025-10-05 7:30 PM ET
    
    FanDuel NFL:
    Chiefs -2.5 (-105) vs Bills +2.5 (-115)  
    Total: O47.5 (-108), U47.5 (-112)
    Updated: Oct 5, 2025 7:35 PM
    
    BetMGM:
    KC Chiefs -3 (-108) vs BUF Bills +3 (-112)
    Over/Under: 47.5 O(-110)/U(-110)
    """

    odds_result = client.extract_odds(raw_odds_data)

    if odds_result["success"]:
        print(f"✅ Extracted {len(odds_result['data']['rows'])} odds entries")
        print(f"   Books found: {odds_result['data']['books_found']}")
        print(f"   Model used: {odds_result['model_used']} ({odds_result['tokens']} tokens)")

        # Show sample extracted data
        for row in odds_result["data"]["rows"][:3]:
            print(f"   • {row['book']}: {row['selection']} @ {row['american_odds']}")
    else:
        print(f"❌ Odds extraction failed: {odds_result['error']}")
        return

    # ✅ Example 2: Build parlays with constraints (gpt-4o)
    print("\n🎯 STEP 2: Build parlays with reasoning (gpt-4o)")
    parlay_result = client.build_parlays(
        odds_data=odds_result["data"]["rows"],
        bankroll=1000,  # $1000 bankroll
        min_ev=0.025,  # 2.5% minimum edge
        max_legs=4,  # Maximum 4-leg parlays
    )

    if parlay_result["success"]:
        parlays = parlay_result["data"]["parlays"]
        print(f"✅ Generated {len(parlays)} profitable parlays")
        print(f"   Model used: {parlay_result['model_used']} ({parlay_result['tokens']} tokens)")

        for i, parlay in enumerate(parlays[:2], 1):  # Show first 2 parlays
            print(f"\n   💰 Parlay #{i} ({parlay['book']}):")
            print(f"      Combined odds: {parlay['combined_odds']}")
            print(f"      Stake: ${parlay['stake_recommendation']:.2f}")
            print(f"      Risk level: {parlay['risk_assessment']['overall_risk']}")
            print(f"      Expected ROI: {parlay.get('expected_roi', 'N/A')}")

            for leg in parlay["legs"]:
                print(f"      • {leg['selection']} @ {leg['odds']}")
    else:
        print(f"❌ Parlay building failed: {parlay_result['error']}")
        return

    # ✅ Example 3: Generate human explanations (gpt-4o-mini)
    print("\n📝 STEP 3: Generate explanations (gpt-4o-mini)")

    if parlays:
        best_parlay = parlays[0]  # Take the first parlay
        explanation_result = client.explain_parlay(best_parlay)

        if explanation_result["success"]:
            print("✅ Generated human explanation:")
            print(
                f"   Model used: {explanation_result['model_used']} ({explanation_result['tokens']} tokens)"
            )
            print(f"\n📋 {explanation_result['explanation']}")
        else:
            print(f"❌ Explanation failed: {explanation_result['error']}")

    # ✅ Example 4: Validate and repair if needed (gpt-4o)
    print("\n🔧 STEP 4: Validate & repair (gpt-4o)")

    # Simulate corrupted data for repair example
    corrupted_parlay = {
        "parlay_id": "invalid_test",
        "book": "unknown_book",  # Invalid book
        "legs": [
            {"game_id": "same_game", "selection": "Team A", "odds": "invalid"},
            {
                "game_id": "same_game",
                "selection": "Team B",
                "odds": "+150",
            },  # Correlation violation
        ],
        "combined_odds": "not_a_number",
    }

    repair_result = client.validate_and_repair(corrupted_parlay)

    if repair_result["success"]:
        print("✅ Validation and repair completed:")
        print(f"   Model used: {repair_result['model_used']} ({repair_result['tokens']} tokens)")
        print(f"   Violations found: {len(repair_result['data']['violations_found'])}")
        print(f"   Repair successful: {repair_result['data']['repair_successful']}")

        if repair_result["data"]["violations_found"]:
            print("   Issues detected:")
            for violation in repair_result["data"]["violations_found"]:
                print(f"   • {violation}")
    else:
        print(f"❌ Validation failed: {repair_result['error']}")

    # ✅ Performance summary
    print("\n📊 PERFORMANCE SUMMARY")
    print("=" * 40)
    total_tokens = sum(
        [
            odds_result.get("tokens", 0),
            parlay_result.get("tokens", 0),
            explanation_result.get("tokens", 0) if "explanation_result" in locals() else 0,
            repair_result.get("tokens", 0),
        ]
    )

    total_time = sum(
        [
            odds_result.get("execution_time", 0),
            parlay_result.get("execution_time", 0),
            explanation_result.get("execution_time", 0) if "explanation_result" in locals() else 0,
            repair_result.get("execution_time", 0),
        ]
    )

    print(f"Total tokens used: {total_tokens:,}")
    print(f"Total execution time: {total_time:.2f}s")
    print("Models used: gpt-4o-mini (extract/explain) + gpt-4o (build/repair)")
    print("\n✅ EQ12 Expert Model Integration - Complete! 🎰")


def advanced_example():
    """Advanced usage patterns for production scenarios."""

    print("\n🚀 ADVANCED USAGE PATTERNS")
    print("=" * 50)

    config = EQ12Config(
        allowed_books=["draftkings", "fanduel", "betmgm"],
        min_ev_threshold=0.04,  # Higher threshold for selectivity
        kelly_cap_per_leg=0.02,  # Conservative Kelly
        max_correlation_risk=0.1,  # Very low correlation
        stale_data_threshold_minutes=10,  # Very fresh data only
    )

    client = EQ12ModelClient(config)

    # ✅ Pattern 1: Batch processing with error handling
    print("\n📦 Batch Processing Example:")

    multiple_sources = [
        "DraftKings: Patriots -7 (-110) vs Jets +7 (-110)",
        "FanDuel: Lakers +5.5 (-108) vs Warriors -5.5 (-112)",
        "BetMGM: Over 225.5 (-110) Under 225.5 (-110) Lakers vs Warriors",
    ]

    batch_results = []
    for i, source in enumerate(multiple_sources, 1):
        print(f"   Processing source {i}...")
        result = client.extract_odds(source)
        batch_results.append(result)

        if result["success"]:
            print(f"   ✅ Extracted {len(result['data']['rows'])} entries")
        else:
            print(f"   ❌ Failed: {result['error']}")

    successful_extractions = [r for r in batch_results if r["success"]]
    print(
        f"\n   📊 Batch complete: {len(successful_extractions)}/{len(multiple_sources)} successful"
    )

    # ✅ Pattern 2: Custom configuration for different risk profiles
    print("\n⚙️  Custom Risk Profile Example:")

    # Conservative profile
    EQ12Config(
        allowed_books=["draftkings"],  # Single book only
        min_ev_threshold=0.05,  # 5% minimum edge
        kelly_cap_per_leg=0.015,  # 1.5% Kelly cap
        max_correlation_risk=0.05,  # Very low correlation
    )

    # Aggressive profile
    EQ12Config(
        allowed_books=["draftkings", "fanduel", "betmgm"],
        min_ev_threshold=0.02,  # 2% minimum edge
        kelly_cap_per_leg=0.03,  # 3% Kelly cap
        max_correlation_risk=0.2,  # Higher correlation tolerance
    )

    print("   Conservative config: Single book, 5% min EV, 1.5% Kelly cap")
    print("   Aggressive config: All books, 2% min EV, 3% Kelly cap")
    print("   → Easily switch between risk profiles!")


if __name__ == "__main__":
    # Run basic examples
    main()

    # Run advanced patterns
    advanced_example()

    print("\n🎯 Examples complete! Check the integration guide for production deployment.")
    print("📁 Next: Integrate with your EQ12 scheduler workflow.")

File: models/examples/usage_examples.ts
/**
 * EQ12 Production Model Client - TypeScript/Node.js Usage Examples
 * Expert-level integration patterns for OpenAI models with EQ12 constraints.
 * 
 * This demonstrates the complete workflow from raw odds → normalized data → parlays
 * using the task-specific model selection from your expert guide.
 */

import EQ12ModelClient, { EQ12Config } from '../eq12_client';

async function main(): Promise<void> {
    console.log('🎰 EQ12 Production Model Client - Node.js Examples');
    console.log('='.repeat(60));
    
    // ✅ Initialize with EQ12-specific constraints
    const config: EQ12Config = {
        allowedBooks: ['draftkings', 'fanduel', 'betmgm'],
        minEvThreshold: 0.03,  // 3% minimum edge
        kellyCapPerLeg: 0.025,  // 2.5% Kelly cap per leg
        maxCorrelationRisk: 0.15,  // Low correlation tolerance
        staleDataThresholdMinutes: 15  // Fresh data only
    };
    
    const client = new EQ12ModelClient(config);
    
    // ✅ Example 1: Extract odds from raw sportsbook data
    console.log('\n📊 STEP 1: Extract & normalize odds (gpt-4o-mini)');
    const rawOddsData = `
    DraftKings NFL Week 5:
    Kansas City Chiefs -3.0 (-110) vs Buffalo Bills +3.0 (-110)
    Total: Over 47.5 (-110), Under 47.5 (-110)
    Last updated: 2025-10-05 7:30 PM ET
    
    FanDuel NFL:
    Chiefs -2.5 (-105) vs Bills +2.5 (-115)  
    Total: O47.5 (-108), U47.5 (-112)
    Updated: Oct 5, 2025 7:35 PM
    
    BetMGM:
    KC Chiefs -3 (-108) vs BUF Bills +3 (-112)
    Over/Under: 47.5 O(-110)/U(-110)
    `;
    
    const oddsResult = await client.extractOdds(rawOddsData);
    
    if (oddsResult.success && oddsResult.data) {
        console.log(`✅ Extracted ${oddsResult.data.rows.length} odds entries`);
        console.log(`   Books found: ${oddsResult.data.booksFound.join(', ')}`);
        console.log(`   Model used: ${oddsResult.modelUsed} (${oddsResult.tokens} tokens)`);
        
        // Show sample extracted data
        oddsResult.data.rows.slice(0, 3).forEach(row => {
            console.log(`   • ${row.book}: ${row.selection} @ ${row.americanOdds}`);
        });
    } else {
        console.log(`❌ Odds extraction failed: ${oddsResult.error}`);
        return;
    }
    
    // ✅ Example 2: Build parlays with constraints (gpt-4o)
    console.log('\n🎯 STEP 2: Build parlays with reasoning (gpt-4o)');
    const parlayResult = await client.buildParlays(
        oddsResult.data!.rows,
        1000,  // $1000 bankroll
        0.025, // 2.5% minimum edge
        4      // Maximum 4-leg parlays
    );
    
    if (parlayResult.success && parlayResult.data) {
        const parlays = parlayResult.data.parlays;
        console.log(`✅ Generated ${parlays.length} profitable parlays`);
        console.log(`   Model used: ${parlayResult.modelUsed} (${parlayResult.tokens} tokens)`);
        
        parlays.slice(0, 2).forEach((parlay, i) => {  // Show first 2 parlays
            console.log(`\n   💰 Parlay #${i + 1} (${parlay.book}):`);
            console.log(`      Combined odds: ${parlay.combinedOdds}`);
            console.log(`      Stake: $${parlay.stakeRecommendation.toFixed(2)}`);
            console.log(`      Risk level: ${parlay.riskAssessment.overallRisk}`);
            
            parlay.legs.forEach(leg => {
                console.log(`      • ${leg.selection} @ ${leg.odds}`);
            });
        });
    } else {
        console.log(`❌ Parlay building failed: ${parlayResult.error}`);
        return;
    }
    
    // ✅ Example 3: Generate human explanations (gpt-4o-mini)  
    console.log('\n📝 STEP 3: Generate explanations (gpt-4o-mini)');
    
    if (parlayResult.data && parlayResult.data.parlays.length > 0) {
        const bestParlay = parlayResult.data.parlays[0];  // Take the first parlay
        const explanationResult = await client.explainParlay(bestParlay);
        
        if (explanationResult.success) {
            console.log('✅ Generated human explanation:');
            console.log(`   Model used: ${explanationResult.modelUsed} (${explanationResult.tokens} tokens)`);
            console.log(`\n📋 ${explanationResult.explanation}`);
        } else {
            console.log(`❌ Explanation failed: ${explanationResult.error}`);
        }
    }
    
    // ✅ Example 4: Validate and repair if needed (gpt-4o)
    console.log('\n🔧 STEP 4: Validate & repair (gpt-4o)');
    
    // Simulate corrupted data for repair example
    const corruptedParlay = {
        parlayId: 'invalid_test',
        book: 'unknown_book',  // Invalid book
        legs: [
            { gameId: 'same_game', selection: 'Team A', odds: 'invalid' as any },
            { gameId: 'same_game', selection: 'Team B', odds: '+150' }  // Correlation violation
        ],
        combinedOdds: 'not_a_number' as any,
        stakeRecommendation: 0,
        riskAssessment: { overallRisk: 'HIGH' as const, correlationRisk: 0, staleDataRisk: false }
    };
    
    const repairResult = await client.validateAndRepair(corruptedParlay);
    
    if (repairResult.success && repairResult.data) {
        console.log('✅ Validation and repair completed:');
        console.log(`   Model used: ${repairResult.modelUsed} (${repairResult.tokens} tokens)`);
        console.log(`   Violations found: ${repairResult.data.violationsFound.length}`);
        console.log(`   Repair successful: ${repairResult.data.repairSuccessful}`);
        
        if (repairResult.data.violationsFound.length > 0) {
            console.log('   Issues detected:');
            repairResult.data.violationsFound.forEach(violation => {
                console.log(`   • ${violation}`);
            });
        }
    } else {
        console.log(`❌ Validation failed: ${repairResult.error}`);
    }
    
    // ✅ Performance summary
    console.log('\n📊 PERFORMANCE SUMMARY');
    console.log('='.repeat(40));
    
    const totalTokens = (oddsResult.tokens || 0) + 
                       (parlayResult.tokens || 0) + 
                       0 + // explanation tokens would go here
                       (repairResult.tokens || 0);
    
    const totalTime = (oddsResult.executionTime || 0) + 
                     (parlayResult.executionTime || 0) + 
                     0 + // explanation time would go here 
                     (repairResult.executionTime || 0);
    
    console.log(`Total tokens used: ${totalTokens.toLocaleString()}`);
    console.log(`Total execution time: ${totalTime.toFixed(2)}s`);
    console.log(`Models used: gpt-4o-mini (extract/explain) + gpt-4o (build/repair)`);
    console.log('\n✅ EQ12 Expert Model Integration - Complete! 🎰');
}

async function advancedExamples(): Promise<void> {
    console.log('\n🚀 ADVANCED USAGE PATTERNS');
    console.log('='.repeat(50));
    
    const config: EQ12Config = {
        allowedBooks: ['draftkings', 'fanduel', 'betmgm'],
        minEvThreshold: 0.04,  // Higher threshold for selectivity
        kellyCapPerLeg: 0.02,  // Conservative Kelly
        maxCorrelationRisk: 0.1,  // Very low correlation
        staleDataThresholdMinutes: 10  // Very fresh data only
    };
    
    const client = new EQ12ModelClient(config);
    
    // ✅ Pattern 1: Batch processing with error handling
    console.log('\n📦 Batch Processing Example:');
    
    const multipleSources = [
        'DraftKings: Patriots -7 (-110) vs Jets +7 (-110)',
        'FanDuel: Lakers +5.5 (-108) vs Warriors -5.5 (-112)', 
        'BetMGM: Over 225.5 (-110) Under 225.5 (-110) Lakers vs Warriors'
    ];
    
    const batchResults = [];
    for (let i = 0; i < multipleSources.length; i++) {
        const source = multipleSources[i];
        console.log(`   Processing source ${i + 1}...`);
        const result = await client.extractOdds(source);
        batchResults.push(result);
        
        if (result.success && result.data) {
            console.log(`   ✅ Extracted ${result.data.rows.length} entries`);
        } else {
            console.log(`   ❌ Failed: ${result.error}`);
        }
    }
    
    const successfulExtractions = batchResults.filter(r => r.success);
    console.log(`\n   📊 Batch complete: ${successfulExtractions.length}/${multipleSources.length} successful`);
    
    // ✅ Pattern 2: Custom configuration for different risk profiles
    console.log('\n⚙️  Custom Risk Profile Example:');
    
    // Conservative profile
    const conservativeConfig: EQ12Config = {
        allowedBooks: ['draftkings'],  // Single book only
        minEvThreshold: 0.05,  // 5% minimum edge
        kellyCapPerLeg: 0.015,  // 1.5% Kelly cap
        maxCorrelationRisk: 0.05  // Very low correlation
    };
    
    // Aggressive profile  
    const aggressiveConfig: EQ12Config = {
        allowedBooks: ['draftkings', 'fanduel', 'betmgm'],
        minEvThreshold: 0.02,  // 2% minimum edge
        kellyCapPerLeg: 0.03,  // 3% Kelly cap
        maxCorrelationRisk: 0.2  // Higher correlation tolerance
    };
    
    console.log('   Conservative config: Single book, 5% min EV, 1.5% Kelly cap');
    console.log('   Aggressive config: All books, 2% min EV, 3% Kelly cap');
    console.log('   → Easily switch between risk profiles!');
}

// Run examples if called directly
if (require.main === module) {
    (async () => {
        try {
            // Run basic examples
            await main();
            
            // Run advanced patterns
            await advancedExamples();
            
            console.log('\n🎯 Examples complete! Check the integration guide for production deployment.');
            console.log('📁 Next: Integrate with your EQ12 scheduler workflow.');
        } catch (error) {
            console.error('❌ Example execution failed:', error);
        }
    })();
}

export { main as runBasicExamples, advancedExamples as runAdvancedExamples };